faq cleanup and additions (#838)
pulumi/pulumi-hugo: themes/default/content/docs/troubleshooting/faq.md

---

> title: FAQ

Author: ```suggestion
title: FAQs
```

### How can I add support for my favorite language?

Supported languages run out of process and communicate over gRPC with the Pulumi engine and resource providers. Check out the [protocol definitions](https://github.com/pulumi/pulumi/tree/master/sdk/proto) along with the language providers themselves. You can look at how we added [support for Go](https://github.com/pulumi/pulumi/pull/1456), which should help with scoping. There is also a summary of the core work items needed as part of adding support for a typical new language on the [New Language wiki page](https://github.com/pulumi/pulumi/wiki/New-Language-Bring-up).

## Automatic Rollbacks

### Does Pulumi support automatic rollback in the event of an error or failure?

No, Pulumi does not automatically rollback changes made during an update if an error or failure occurs. In the event that an error or failure occurs, Pulumi will complete any pending operations
currently in progress and then exit and report the error or failure.

To accomplish a _manual_ rollback after a failed deployment, revert the code and configuration changes of the failed deployment and run `pulumi up` to update your infrastructure to its previous state.

> There is a [GitHub issue related to the idea of automatic rollbacks](https://github.com/pulumi/pulumi/issues/96) if you would like to add to the discussion.

Author: ive added this to the issue instead so we don't have to maintain this if that process adjusts or changes in the future

## Organizations

## How do I create a stack inside an Organization instead of my User account?

To create a stack in a different Pulumi organization, prefix the stack's
name with the organization name. For example:

```sh
$ pulumi stack init acme-corp/widget-server
```

## Pulumi Service

## How does Pulumi store state?

> Pulumi needs to store the result of operations. On creation of a Pulumi resource, Pulumi makes a call to the cloud providers’ API and then it stores the result of that API call. The place where Pulumi stores that result is called the “state”. The state can be stored using the Pulumi service, Amazon S3, Azure Blob Storage, or Google Cloud Storage Buckets.

Author: this is one of the questions added from Lee's blog post

> Pulumi needs to store the result of operations. On creation of a Pulumi resource, Pulumi makes a call to the cloud providers’ API and then it stores the result of that API call. The place where Pulumi stores that result is called the “state”. The state can be stored using the Pulumi service, Amazon S3, Azure Blob Storage, or Google Cloud Storage Buckets.

Reviewer: Because we also refer to it as a checkpoint: 

```suggestion
Pulumi needs to store the result of operations. On creation of a Pulumi resource, Pulumi makes a call to the cloud provider's API and then it stores the result of that API call. The place where Pulumi stores that result is called the "state" or "checkpoint". The state can be stored using the Pulumi Service, Amazon S3, Azure Blob Storage, Google Cloud Storage Buckets, or as a file on your local machine that you manage yourself.
```

> Pulumi needs to store the result of operations. On creation of a Pulumi resource, Pulumi makes a call to the cloud providers’ API and then it stores the result of that API call. The place where Pulumi stores that result is called the “state”. The state can be stored using the Pulumi service, Amazon S3, Azure Blob Storage, or Google Cloud Storage Buckets.

Author: ive taken some of your suggestions and some from @cnunciato

### What happens if app.pulumi.com is down?

Any infrastructure that you’ve deployed using Pulumi will continue working and can be managed with your cloud provider’s console or CLI, that is, app.pulumi.com should not affect any runtime behavior of your application.

If app.pulumi.com is down, you'll be unable to preview, update or destroy a stack using Pulumi.  Some commands, like `pulumi logs`, use app.pulumi.com to find the correct log stream, so will not function until pulumi.com recovers; however, your cloud provider will still produce logs that you can use for diagnostics and you can view these via your cloud console or CLI.

### Can I use Pulumi without depending on the Pulumi Service?

> We think that using the Pulumi service and the Pulumi tool together provide the right combination of usability, safety, and security. However, for users with especially unique requirements, it is possible to use the Pulumi tool apart from the Service.

Author: ive fixed this, but this exists today, i did not write this.

When you use Pulumi without the Pulumi Service, the checkpoint for your stack is stored locally or in your own external self-managed state storage. If that file is lost or outdated, Pulumi can no longer operate on your stack. To collaborate with others on your stack, you must host this file yourself and protect against conflicting updates to it. If you use your own checkpoint file, the Pulumi Service features, such as the deployment history and resource view, will not be available.

To use Pulumi without the Pulumi Service, log in using `pulumi login --local` or by logging in to an alternative backend. For more information, read more at [State and Backends]({{< relref "/docs/intro/concepts/state" >}}).

### How can I go back to using the Pulumi Service?

Run `pulumi login` and you’ll be back to using the Pulumi Service. You will need to migrate any existing stacks to the Service.

> ### I've been using the local endpoint, can I migrate to the Pulumi Service?

Author: ```suggestion
### I've been using a local backend, can I migrate to the Pulumi Service?
```

## Resource names

### Why do resource names have random hex character suffixes?

Pulumi resources have different logical names and physical names. The logical name is how your resource is known to Pulumi, and is part of the auto-assigned resource URN, while the physical name is how your resource is known to your cloud provider. The logical name only needs to be unique within a stack, while the physical name typically needs to be unique within a certain scope (sometimes even globally).

To ensure that as many logical Pulumi resources and stacks can be stood up side-by-side without them conflicting, Pulumi automatically creates each resource's name by combining the logical name with a random suffix appended afterwards. This approach to auto-naming also ensures Pulumi can replace resources with less downtime, by creating the replacement before needing to delete the old resource. In practice, both of these provide significant benefits. This means, however, that your resource's physical name will typically look like `my-bucket-d7c2fa0` instead of `my-bucket`.

> You can disable this [auto-naming]({{< relref "programming-model" >}}#autonaming) on a per-resource basis if this isn't right for you. Doing so lets you have precise physical names without random suffixes, but also means you lose the two benefits named above.

Reviewer: ```suggestion
You can disable this [auto-naming]({{< relref "docs/intro/concepts/resources/#autonaming" >}}#autonaming) on a per-resource basis if this isn't right for you. Doing so lets you have precise physical names without random suffixes, but also means you lose the two benefits named above.
```

## Secrets

### How does Pulumi manage secrets?

When you set a [configuration]({{< relref "/docs/intro/concepts/config" >}}) value, you may pass `--secret` to `pulumi config set` which causes the value to be encrypted so it can be safely persisted in `Pulumi.<stack-name>.yaml`. For every stack the Pulumi service manages a unique encryption key, which it uses to encrypt secrets for that stack (and this is configurable to use your own custom secrets provider). Because a different key is used for each stack, encrypting the same value across two different stacks will lead to different encrypted strings being stored in the `Pulumi.<stack-name>.yaml` files. This also means that you cannot copy an encrypted value from one file to another using a text editor. Instead, you must use `pulumi config set`.

When you run a preview, update or destroy, pulumi decrypts this data. It is plain text during the execution of your deployment, and any part of your Pulumi program may access it using the Pulumi config object.

### Are my secrets ever visible?

> Pulumi provides primitives so you can enforce your [secrets]({{< relref "/docs/intro/concepts/secrets#secrets" >}}) are stored in a secure manner in the CLI, state file and Pulumi Console. During an update, your secrets will be unencrypted in memory and visible to your Pulumi program. It is your responsibility to ensure that you do not persist them outside of Pulumi without securing them.

Reviewer: Not your PR, but my IDE caught it as a typo. 😄 (And "decrypt" is also the term we use for this operation.) 

```suggestion
Pulumi provides primitives so you can enforce your [secrets]({{< relref "/docs/intro/concepts/secrets#secrets" >}}) are stored in a secure manner in the CLI, state file and Pulumi Console. During an update, your secrets will be decrypted in memory and visible to your Pulumi program. It is your responsibility to ensure that you do not persist them outside of Pulumi without securing them.
```

## Understanding Pulumi

### Does Pulumi use Terraform?

> Pulumi does not use the Terraform engine in any way. However, some Pulumi providers leverage the Terraform schemas in order to know what resources and parameters are available, and to determine the return and response attributes. On the other hand, Pulumi native providers are mapped directly to the cloud provider APIs and do not leverage the Terraform schemas.

Reviewer: This section feels like it could use a few links, since not all users will know what things like "Terraform schemas" and "native providers" are, and they seem to be relevant to the answer. What do you think of maybe adding these for clarification (and more actively pointing folks to native providers)?

```suggestion
Pulumi does not use the Terraform engine in any way. However, some Pulumi providers use [Terraform schemas](https://www.terraform.io/docs/extend/schemas/index.html) in order to know what resources and parameters are available, and to determine the return and response attributes. On the other hand, [Pulumi native providers](https://www.pulumi.com/blog/pulumiup-native-providers/) are mapped directly to the cloud provider APIs and do not use Terraform schemas.
```

> Pulumi does not use the Terraform engine in any way. However, some Pulumi providers leverage the Terraform schemas in order to know what resources and parameters are available, and to determine the return and response attributes. On the other hand, Pulumi native providers are mapped directly to the cloud provider APIs and do not leverage the Terraform schemas.

Reviewer 2: ```suggestion
Pulumi does not use the Terraform engine in any way. However, some Pulumi providers leverage the open-source Terraform schemas to know which resources and parameters are available and to determine the return and response attributes. On the other hand, Pulumi native providers are mapped directly to the cloud provider APIs and do not leverage the Terraform schemas.
```

### Is Pulumi imperative or declaritive?

Pulumi is a declarative tool that uses imperative languages to define your end state. The language is used for authoring your program, it’s not used for talking to the cloud provider API.

### How will Pulumi make me more productive?

> Pulumi uses strongly typed languages with programming languages that support [Intellisense](https://code.visualstudio.com/docs/editor/intellisense) and the [Language Server Protocol](https://en.wikipedia.org/wiki/Language_Server_Protocol) which means when you are defining Pulumi programs you rarely need to leave your IDE.

Reviewer: ```suggestion
Pulumi uses strongly typed languages with programming languages that support [IntelliSense](https://code.visualstudio.com/docs/editor/intellisense) and the [Language Server Protocol](https://en.wikipedia.org/wiki/Language_Server_Protocol) which means when you are defining Pulumi programs you rarely need to leave your IDE.
```

When you run a preview, update or destroy, pulumi decrypts this data. It is plain text during the execution of your deployment, and any part of your Pulumi program may access it using the Pulumi config object.

### Are my secrets ever visible?

Pulumi provides primitives so you can enforce your [secrets]({{< relref "/docs/intro/concepts/secrets#secrets" >}}) are stored in a secure manner in the CLI, state file and Pulumi Console. During an update, your secrets will be unencrypted in memory and visible to your Pulumi program. It is your responsibility to ensure that you do not persist them outside of Pulumi without securing them.

> ## Understanding Pulumi

Author: idk if this is the right heading, anyone have better ideas?

Reviewer 2: I think it's a good heading, personally.

### Does Pulumi use Terraform?

Some Pulumi providers leverage the Terraform schemas in order to know what resources and parameters are available, and to determine the return and response attributes.

> Pulumi does also provide native providers which are mapped directly to the cloud provider APIs and do not leverage the Terraform schemas.

Author: this is one of the questions added from Lee's blog post

Reviewer 3: I like that we added this question. I wonder if we should be even a bit more verbose, e.g. 
```suggestion
### Does Pulumi use Terraform?

Pulumi does not use the Terraform engine in any way. However, some Pulumi providers leverage the Terraform schemas in order to know what resources and parameters are available, and to determine the return and response attributes. On the other hand, Pulumi native providers are mapped directly to the cloud provider APIs and do not leverage the Terraform schemas.
```

### Is Pulumi imperative or declaritive?

> Pulumi is a declarative tool that uses imperative languages to define your end state. The language is used for authoring your program, it’s not used for talking to the cloud provider API.

Author: this is one of the questions added from Lee's blog post

Reviewer 2: ```suggestion
### Is Pulumi imperative or declaritive?

Pulumi is a declarative tool that uses imperative languages to define your end state. The language is used for authoring your program; it’s not used for talking to the cloud provider API.
```

### How will Pulumi make me more productive?

> Pulumi uses strongly typed languages with programming languages that support [Intellisense](https://code.visualstudio.com/docs/editor/intellisense) and the [Language Server Protocol](https://en.wikipedia.org/wiki/Language_Server_Protocol) which means when you are defining Pulumi programs you rarely need to leave your IDE.

Author: this is one of the questions added from Lee's blog post

Reviewer 2: ```suggestion
### How will Pulumi make me more productive?

Pulumi uses strongly typed languages with programming languages that support [Intellisense](https://code.visualstudio.com/docs/editor/intellisense) and the [Language Server Protocol](https://en.wikipedia.org/wiki/Language_Server_Protocol) among other standard IDE tooling, which means when you are defining Pulumi programs, you rarely need to leave your IDE.
```
